components/builder-jobsrv/src/server/log_ingester.rs

```diff
@@ -0,0 +1,142 @@
+            //    L = a line of log output
+            //    C = the log is complete
+            // 3: a protobuf message
+
+            self.intake_sock.recv(&mut self.msg, 0)?; // identity frame
+
+            match str::from_utf8(self.intake_sock.recv_bytes(0).unwrap().as_slice()).unwrap() {
+                "L" => {
+                    self.intake_sock.recv(&mut self.msg, 0)?; // protobuf message frame
+                    match parse_from_bytes::<JobLogChunk>(&self.msg) {
+                        Ok(chunk) => {
+                            let log_file = self.log_dir.log_file_path(chunk.get_job_id());
+
+                            // TODO: Consider caching file handles for
+                            // currently-processing logs.
+                            let open = OpenOptions::new()
+                                .create(true)
+                                .append(true)
+                                .open(log_file.as_path());
+
+                            match open {
+                                Ok(mut file) => {
+                                    // TODO: Just write the chunk body; line number is
+                                    // for debugging only
+                                    file.write(format!("{}: {}",
+                                                       chunk.get_seq(),
+                                                       chunk.get_body()).as_bytes())?;
+                                    file.flush()?;
+                                    println!("LOG: {:?}: {:?}",
+                                             chunk.get_seq(),
+                                             chunk.get_body());
+                                },
+                                Err(e) => {
+                                    warn!("Could not open {:?} for appending! {:?}", log_file, e);
+                                }
+                            }
+                        },
+                        Err(e) => {
+                            warn!("ERROR parsing JobLogChunk: {:?}", e);
+                        }
+                    }
+                },
+                "C" => {
+                    self.intake_sock.recv(&mut self.msg, 0)?; // protobuf message frame
+                    match parse_from_bytes::<JobLogComplete>(&self.msg) {
+                        Ok(complete) => {
+                            let id = complete.get_job_id();
+                            debug!("Log complete for job {:?}", id);
+
+                            // TODO: Save logs in long-term storage,
+                            // like S3
+                            let log_file = self.log_dir.log_file_path(id);
+
+                            // Until we can ship things off to an
+                            // S3-alike, we'll just use a file URL
+                            let url = format!("file://{}", log_file.to_str().unwrap());
+                            self.data_store.set_log_url(id, &url)?;
```

Comment:
Right, the S3/local fs logic is fine.. the confusing thing (to me) was the job_log_get does not use the log url, though it is getting set to a file://.. url.  If we say, the location is always in the log_url (whether the location points to a local fs or a remote location), then that seems more consistent for at least the first time reader of the code. The way you have it is an extra special case to parse mentally. Not a big deal though, I wouldn't hold off the merge or anything.